src/cdomains/intDomain.ml

```diff
@@ -349,6 +349,446 @@ struct
+    else Some (SS.add (x,y) SS.empty)
+
+  let norm_interval (x, y) =
+    if Int64.compare min_int x > 0 || Int64.compare max_int y < 0 then (min_int, max_int)
+    else (x, y)
+
+  let (@@) f x = f x
+
+  let to_int = function
+    | Some s when SS.cardinal s = 1 ->
+        let (x, y) = SS.min_elt s in
+        if Int64.compare x y = 0 then Some x else None
+    | _ -> None
+
+  let of_int x =
+    let s = SS.empty in
+    let s_of_int = SS.add (x, x) s in
+    Some s_of_int
+
+  let is_int = function
+    | Some s when SS.cardinal s = 1 ->
+        let x, y = SS.min_elt s in
+        if Int64.compare x y = 0 then true else false
+    | _ -> false
+
+  let equal_to i = function
+    | None -> failwith "unsupported: equal_to with bottom"
+    | Some s ->
+      if not (SS.exists (fun (a, b) -> a <= i && i <= b) s) then `Neq else
+        match SS.cardinal s with
+          | 1 ->
+            let a, b = SS.min_elt s in
+            if a = b && b = i then `Eq else `Top
+          | _ -> `Top
+
+  let to_bool = function
+    | None -> None
+    | x when equal x zero -> Some false
+    | x -> if leq zero x then None else Some true
+
+  let to_bool_interval x =
+    match x with
+    | None -> None
+    | _ when equal x zero -> x
+    | _ -> if leq zero x then top_bool else one
+
+  let of_bool = function
+    | true -> Some (SS.add (1L, 1L) SS.empty)
+    | false -> Some (SS.add (0L, 0L) SS.empty)
+
+  let is_bool x =
+    x <> None && not (leq zero x) || equal x zero
+
+  let to_excl_list _ = None
+  let of_excl_list t _ = top ()
+  let is_excl_list _ = false
+
+  let of_interval (x,y) = norm @@ Some (x,y)
+  let of_int x = of_interval (x,x)
+  let zero = Some (SS.add (0L, 0L) SS.empty)
+  let one = Some (SS.add (1L, 1L) SS.empty)
+  let top_bool = Some (SS.add (0L, 1L) SS.empty)
+
+  let starting n = norm @@ Some (n, max_int)
+  let ending   n = norm @@ Some (min_int, n)
+  let maximal = function
+    | None -> None
+    | Some s ->
+      let max_elt = SS.max_elt s in
+      Some ((fun (x, y) -> y) max_elt)
+  let minimal = function
+    | None -> None
+    | Some s ->
+      let min_elt = SS.min_elt s in
+      Some ((fun (x, y) -> x) min_elt)
+
+  (* takes a sorted list of intervals, iterates through them and merges two adjacent if they intersect*)
+  let rec merge_intersect ?(acc = []) l =
+    match l with
+    | (x1, x2) :: xs when Int64.compare x1 x2 > 0 ->
+      merge_intersect ~acc:acc xs
+    | (x1, x2) :: (y1, y2) :: xs ->
+      if Int64.compare x2 (Int64.sub y1 1L) >= 0 then
+        merge_intersect ~acc:acc ((x1, max x2 y2) :: xs)
+      else merge_intersect ~acc:((x1, x2) :: acc) ((y1, y2) :: xs)
+    | [(x1, x2)] -> merge_intersect ~acc:((x1, x2) :: acc) []
+    | [] -> acc
+
+  (* recursive template for evaluating arithmetic operations which are performed pairwise
+     for each interval from both set operands *)
+  let rec arithm ?(acc1 = []) ?(acc2 = []) f l1 l2 =
```

Comment:
Probably you would need to somehow normalize the results of these operations?
So that e.g. we do not have the same interval in there, or intervals that are overlapping etc.

Reply:
Yes, it is already normalized; I double-checked the `merge_intersect` function, it works as intended.